packages/flutter/lib/src/services/binding.dart

```diff
@@ -263,30 +264,110 @@ mixin ServicesBinding on BindingBase, SchedulerBinding {
     return null;
   }
 
-  Future<void> _handlePlatformMessage(MethodCall methodCall) async {
+  Future<dynamic> _handlePlatformMessage(MethodCall methodCall) async {
     final String method = methodCall.method;
-    // There is only one incoming method call currently possible.
-    assert(method == 'SystemChrome.systemUIChange');
-    final List<dynamic> args = methodCall.arguments as List<dynamic>;
-    if (_systemUiChangeCallback != null) {
-      await _systemUiChangeCallback!(args[0] as bool);
+    assert(method == 'SystemChrome.systemUIChange' || method == 'System.requestAppExit');
+    switch (method) {
+      case 'SystemChrome.systemUIChange':
+        final List<dynamic> args = methodCall.arguments as List<dynamic>;
+        if (_systemUiChangeCallback != null) {
+          await _systemUiChangeCallback!(args[0] as bool);
+        }
+        break;
+      case 'System.requestAppExit':
+        return <String, dynamic>{'response': (await handleRequestAppExit()).name};
     }
   }
 
   static AppLifecycleState? _parseAppLifecycleMessage(String message) {
     switch (message) {
-      case 'AppLifecycleState.paused':
-        return AppLifecycleState.paused;
       case 'AppLifecycleState.resumed':
         return AppLifecycleState.resumed;
       case 'AppLifecycleState.inactive':
         return AppLifecycleState.inactive;
+      case 'AppLifecycleState.paused':
+        return AppLifecycleState.paused;
       case 'AppLifecycleState.detached':
         return AppLifecycleState.detached;
     }
     return null;
   }
 
+  /// Handles any requests for application exit that may be received on the
+  /// [SystemChannels.platform] method channel.
+  ///
```

Comment:
Should we document that there is no guarantee that this method will be called before the app exists since there are ways to exit the app without leaving the option to intervene. In other words, apps should not rely on this to e.g. save state before the app quits or anything like that.